Tidy up readme (#58)
UMM-CSci-3601/intro-to-git: README.md

the same place when that happens.

### Resources for learning `git` and friends

This lab experience definitely will _not_ provide a comprehensive overview of the
_many_ features `git` provides, and there are a _lot_ of on-line resources
that can provide additional information.

Some that you might want to look at right away
to help you get up to speed on `git` and its
friends include:

* A few of the _many_ GitHub tutorials and
  docs. There are a _lot_ more than just these, but these are short and pretty easy to spin through, and cover the key concepts we’ll need at the beginning of the semester.
  * [GitHub Hello World](https://guides.github.com/activities/hello-world/) (a really nice,
    short introduction to the key ideas)
  * [Getting started with GitHub](https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github)
  * [Understanding the GitHub flow](https://guides.github.com/introduction/flow/)
  * [Mastering Markdown](https://guides.github.com/features/mastering-markdown/)
    (Markdown is the markup language used
    by Slack, GitHub, and _many_ other
    online tools, so understanding at least

>     the basics is really valuable.)

Reviewer: Should this be on a sub-bullet like the one below is?

Author: Yup, it almost certainly should. Thanks!

Before we actually start to _use_ `git`, you should [configure your `git`
email so work you do in our lab properly connects to your GitHub
account](https://help.github.com/articles/setting-your-commit-email-address-in-git/).
You only need to do this once and it will "stick" throughout the course
(and beyond) in all systems that use `git`.

* Open a terminal window

>   * :mask: On Macs and Linux boxes use the Terminal app. On Windows use Powershell.

Reviewer: ```suggestion
  * :mask: On Macs and Linux boxes use the Terminal app. On Windows use PowerShell.
```

### Configuring your `git` e-mail

:mask::warning::exclamation::warning::mask:

> This whole section may be irrelevant for home set ups. It looks like

Reviewer: ```suggestion
This whole section may be irrelevant for home set-ups. It looks like
```

#### Have a look at the project

Open up `src -> main/java/hellos -> Hellos.java` and you should see a small
Java program. Now open up `src -> test/java/hellos -> HellosTest.java` and you
should see our JUnit tests.

> [JUnit is a widely used testing framework for Java](https://junit.org/junit5/docs/current/user-guide/#writing-tests).
> A Junit test is a method that has the special `@Test` annotation and contains
> one or more _assertions_ like `assertTrue` or `assertEquals`. Tests _pass_
> if all the assertions in a test method are true; they _fail_ if any assertion
> is false.

As well as running tests from an "external" terminal window, you can also run
them from a terminal inside VS Code. Go to the "Terminal" menu and select
"New Terminal". That should bring up a terminal along the bottom of your VS Code
window. That terminal is just like an "external" terminal and you can use it
to do anything you could have done in an "external" terminal window. This
includes things like:

```bash

>    ./gradlew test

Reviewer: ```suggestion
./gradlew test
```

Author: I actually indented it on purpose as I think it reads a little better indented. I guess is could be a source of confusion, but that doesn't seem a huge risk?

Reviewer: It does seem odd for that one to be different from all the other instances in the doc.

It also tells you how you can _unstage_ a file that you might have staged by
accident.

**Commit the change** _to your local copy of the repository_ with

```bash
git commit
```

> * This will bring up a (command line) editor for you to enter

Reviewer: The editor may not be a command line one. For students on Windows we have even suggested they change the setting when installing Git to VS Code:

![image](https://user-images.githubusercontent.com/1300395/104276747-1d9d1600-546b-11eb-98f3-0a45777ac4d7.png)

(https://gist.github.com/floogulinc/2d9f80b278332b0df731e777c43b206c#install-git)

Though vim is the default when you run through the installer and we could update the setup guide to suggest this. 

![image](https://user-images.githubusercontent.com/1300395/104276941-808ead00-546b-11eb-9c0f-eca69115fc16.png)

Author: Thanks for that. I'm fine with them using VS Code as their editor, so I'll just update my text.

working in other branches or on the build system, so everything will be
nice.

Every new `git` repository has one default branch called `main`; this is
usually where the current "deployed" or "released" version of the project
lives, while active development, bug fixes, and the like happen in other
development branches that are merged into `main` when that work is deemed
"done done" (passes the tests, has been through a code review, etc.).

> Historically the default has actually been called `master`. One side-effect
> of the Black Lives Matter movement has been an increasing recognition of
> the painful connotations that term has for many people, and there have
> been [calls to change from `master` to the more neutral `main`](https://dev.to/afrodevgirl/replacing-master-with-main-in-github-2fjf).

> > GitHub is in fact working to change the default naming for new projects,

Author: I assume that copies made from templates (what GitHub Classroom does) will follow the existing branch names (it really has to), so we still have to change our branch names in our repos.

I actually meant to link to the official page, and obviously gapped it – thanks for catching that.

Here, if Chris clicks the "Resolve conflicts" button, GitHub will put Chris in
a (web) editor where they can resolve the conflicts. That editor will likely
have a section that looks something like:

```java
<<<<<<< chris-greeting
    builder.append(chrisSaysHello());
=======
    builder.append(patSaysHello());
>>>>>>> main
```

What this rather odd syntax says is that Chris's `chris-greeting` branch has
a line that calls `chrisSaysHello()`, where `main` (which has Pat's changes)
has a line that calls `patSaysHello()`. Everything between the `<<<<` line
and the `====` line is what things look like in the `chris-greeting` branch,
and everything between the `====` line and the `>>>>` line are what things
look like in the `main` branch.

There are numerous ways this conflict could have come about, such as:

* Chris wants to change the line from calling `patSaysHello()` to
  instead called the (supposedly different) method `chrisSaysHello()`.
* Chris has renamed `patSaysHello()` to `chrisSaysHello()`, and wants to change
  the call to match this renaming.
* Chris wants to make _both_ function calls (in some order, which `git` couldn't
  begin to guess).

> Here `git can't tell which of these happened, and thus can't "know" how to

Reviewer 2: ```suggestion
Here `git` can't tell which of these happened, and thus can't "know" how to
```

Looks like you're missing an accent, by the way! :)

> handle the merge. Thus it punts it to us, the human users to sort out. Sometimes

Reviewer 2: ```suggestion
handle the merge. Thus it punts it to us, the human users, to sort out. Sometimes
```
(also a comma)

and `>>>>` lines, leaving the two calls in the desired order. When things look good,
Chris clicks "Mark as resolved", and then the button to commit the merged changes.

:warning: This is a _very_ simple example of a merge conflict. They can be much more
complex than this, and trickier to resolve. The Pro version of GitKraken (which you
got through the GitHub Student Pack) actually has a very nice GUI for resolving merge
conflicts, and you'd probably rather use that for more complex conflicts.

:warning: :bangbang: At this point Chris (or you if you've been doing something
similar) will need to pull these changes back down to your copy if you want
or need to make changes to your branch. If, for example, Chris realizes here
that the tests are failing because they have the alphabetical order wrong,
they'll need to pull down these changes, fix the alphabetical order, push back
up, and attempt to merge again.

##### You need a review

Now you've finally gotten your code up-to-date and all the checks pass. You
still can't merge because of a big red X and the message "Review required".
This is because we're following a common "best practice" and requiring at

> least one code review _by another person_ for each pull request.

Reviewer: Are we changing this to two code reviews? If so:
```suggestion
least two code review _by other people_ for each pull request.
```

Author: Good catch – thanks!

So you need to get someone to review your code! In the upper right there's
a "Reviewers" section. Click on the gear there and you should get a drop down
with the (GitHub) names of the other people in the class. Choose two or three

> and click their names to invite them to review your code. _Don't just choose
> the top three – if everyone does that the same few people will be asked to do
> _all_ the reviews. Instead, try choosing the two or three names right after yours
> on the list – that should spread the load around.

Reviewer 2: It looks like this `_` is unclosed. I can't tell what your intention was here, but as a suggestion, maybe we want to leave this sentence un-italicized?

```suggestion
and click their names to invite them to review your code. Don't just choose
the top three – if everyone does that the same few people will be asked to do
_all_ the reviews. Instead, try choosing the two or three names right after yours
on the list – that should spread the load around.
```
